### PR TITLE
Implement privilege token for mfa management in UI dashboard

### PR DESCRIPTION
**Pull request by kimlisa**
backend of https://github.com/gravitational/teleport/issues/7773

#### Description
Allow users to view and manage their mfa devices in their logged in UI dashboard. Users are required to acquire a privilege token before adding and deleting a device. 

- Created a new user token subkind called `privilege`. This 5 min short lived token type is used for privileged actions that require users to re-authenticate with their second factor while logged in. Reuses existing `user lock` that when re-auth fails some number of times, it locks the user and prevents them from trying again.
- Added new endpoints that allow users to
  - add mfa device with a privilege token
  - get their own mfa challenges to re-auth with their u2f
  - get list of their own mfa devices
  - get their own privilege token after successful re-auth

### CODE REVIEW

File: lib/web/apiserver.go [hunk 1]
```diff
@@ -349,12 +351,15 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*RewritingHandler, error) {
 	h.POST("/webapi/u2f/password/changerequest", h.WithAuth(h.u2fChangePasswordRequest))
 	h.POST("/webapi/u2f/signrequest", httplib.MakeHandler(h.mfaChallengeRequest))
 	h.POST("/webapi/u2f/signrequest/token", httplib.MakeHandler(h.getMFAChallengeRequestWithTokenHandle))
+	h.POST("/webapi/u2f/signrequest/auth", h.WithAuth(h.getMFAChallengeRequestWithAuthHandle))
```
Comment (kimlisa) [inline lib/web/apiserver.go:354]: Had trouble naming this, any suggestions welcome. There are 3 types of `u2f/signrequest`:

`h.POST("/webapi/u2f/signrequest", httplib.MakeHandler(h.mfaChallengeRequest))`: public handler where auth is their username and password during login

`h.POST("/webapi/u2f/signrequest/token", httplib.MakeHandler(h.getMFAChallengeRequestWithTokenHandle))`: public handler where auth is the token, used during account recovery

`h.POST("/webapi/u2f/signrequest/auth", h.WithAuth(h.getMFAChallengeRequestWithAuthHandle))`: private handler where auth is that the requestor == currently logged in user

File: api/client/client.go [hunk 2]
```diff
@@ -1137,6 +1147,16 @@ func (c *Client) GetMFADevices(ctx context.Context, in *proto.GetMFADevicesReque
 	return resp, nil
 }
 
+// GetMFAAuthenticateChallengeWithAuth gets mfa challenges for the currently logged in user.
```
Comment (zmb3) [inline api/client/client.go:1150]:
```suggestion
// GetMFAAuthenticateChallengeWithAuth gets MFA challenges for the currently logged in user.
```

MFA is an acronym, so let's capitalize it in the godocs.

File: api/client/proto/authservice.proto [hunk 3]
```diff
@@ -802,13 +802,33 @@ message DeleteMFADeviceWithTokenRequest {
...
+// AddMFADeviceWithTokenRequest is a request to add a mfa device for the user defined in
+// token.
+message AddMFADeviceWithTokenRequest {
+    // TokenID is the ID of a user token.
+    string TokenID = 1;
+    // DeviceName is the name of the device.
+    string DeviceName = 2;
+    // SecondFactorCred is the new mfa device auth credentials.
+    oneof SecondFactorCred {
+        // SecondFactorToken is an totp value.
```
Comment (zmb3) [inline api/client/proto/authservice.proto:814]:
```suggestion
        // SecondFactorToken is a TOTP value.
```

File: api/client/proto/authservice.proto [hunk 4]
```diff
@@ -1379,11 +1415,17 @@ service AuthService {
         returns (CreateRecoveryCodesWithTokenResponse);
 
     // GetMFAAuthenticateChallengeWithToken retrieves challenges for all mfa devices for the user
```
Comment (zmb3) [inline api/client/proto/authservice.proto:1417]:
```suggestion
    // GetMFAAuthenticateChallengeWithAuth retrieves challenges for all MFA devices for the user
```

File: lib/web/apiserver.go (hunk 1 above)
Comment (codingllama) [reply lib/web/apiserver.go:354]: I'll review properly later, but in relation to this we've got an ongoing RFD where I'm aliasing (and eventually moving) /u2f/ endpoints to more generic /mfa/ counterparts, which is applicable here too.

Please have a look at the Login and Registration API sections and let me know what you think: [WebAuthn RFD](https://github.com/gravitational/teleport/pull/7808).

File: lib/auth/auth.go [hunk 5]
```diff
@@ -1173,15 +1183,16 @@ func (a *Server) GetMFADevicesWithToken(ctx context.Context, req *proto.GetMFADe
...
 func (a *Server) DeleteMFADeviceWithToken(ctx context.Context, req *proto.DeleteMFADeviceWithTokenRequest) error {
 	token, err := a.GetUserToken(ctx, req.GetTokenID())
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
 	// Restrict to certain kinds of user token and usage.
-	if token.GetSubKind() != UserTokenTypeRecoveryApproved || token.GetUsage() != types.UserTokenUsage_RECOVER_2FA {
+	not2FAApprovedTokenType := token.GetSubKind() != UserTokenTypeRecoveryApproved || token.GetUsage() != types.UserTokenUsage_RECOVER_2FA
+	if not2FAApprovedTokenType && token.GetSubKind() != UserTokenTypePrivilege {
```
Comment (zmb3) [inline lib/auth/auth.go:1195]: This conditional is getting a little bit hard to follow. I would either expand in the comments or even refactor it into a helper method.

File: lib/auth/auth.go [hunk 6]
```diff
@@ -1214,6 +1225,92 @@ func (a *Server) DeleteMFADeviceWithToken(ctx context.Context, req *proto.Delete
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	secondFactor := authPref.GetSecondFactor()
+	var newDevice *types.MFADevice
+
+	switch req.GetSecondFactorCred().(type) {
+	case *proto.AddMFADeviceWithTokenRequest_U2FRegisterResponse:
+		if secondFactor == constants.SecondFactorOTP {
```
Comment (zmb3) [inline lib/auth/auth.go:1254]: Is this if statement correct? I'm confused by the error that says "cluster only allows OTP" because the second factor in this case is OTP. Same with U2F below.

I think the bad parameter error should say "cluster only allows U2F for second factor" - right?

**Review by zmb3 — COMMENTED** [review]
Looks good, just a couple minor suggestions.

File: api/client/client.go [hunk 7]
```diff
@@ -681,6 +681,16 @@ func (c *Client) CreateRecoveryStartToken(ctx context.Context, req *proto.Create
...
 }
 
+// CreatePrivilegeToken returns a new privilege token after user successfully re-auth with their second factor.
+func (c *Client) CreatePrivilegeToken(ctx context.Context, req *proto.CreatePrivilegeTokenRequest) (types.UserToken, error) {
+	token, err := c.grpc.CreatePrivilegeToken(ctx, req, c.callOpts...)
+	if err != nil {
+		return nil, trail.FromGRPC(err)
+	}
+
+	return token, nil
```
Comment (codingllama) [inline api/client/client.go:691]:
```suggestion
	return token, trail.FromGRPC(err)
```

File: api/client/proto/authservice.proto [hunk 8]
```diff
@@ -802,13 +802,33 @@ message DeleteMFADeviceWithTokenRequest {
...
 
+// AddMFADeviceWithTokenRequest is a request to add a mfa device for the user defined in
+// token.
+message AddMFADeviceWithTokenRequest {
+    // TokenID is the ID of a user token.
+    string TokenID = 1;
+    // DeviceName is the name of the device.
+    string DeviceName = 2;
+    // SecondFactorCred is the new mfa device auth credentials.
+    oneof SecondFactorCred {
```
Comment (codingllama) [inline api/client/proto/authservice.proto:813]: Let's use MFARegisterResponse here, so it plays nicer with existing API and ongoing MFA changes.

File: api/client/proto/authservice.proto [hunk 9]
```diff
@@ -1053,6 +1073,19 @@ message CreateRecoveryCodesWithTokenResponse {
     repeated string RecoveryCodes = 1;
 }
 
+// CreatePrivilegeTokenRequest defines a request to create and get a user privilege token.
+message CreatePrivilegeTokenRequest {
+    // Username is the name of user requesting token.
+    string Username = 1;
+    // SecondFactorCred is one of users second factor auth credential.
+    oneof SecondFactorCred {
```
Comment (codingllama) [inline api/client/proto/authservice.proto:1081]: Similarly as above, let's plug MFAAuthenticateResponse here instead.

File: api/client/proto/authservice.proto [hunk 10]
```diff
@@ -1379,11 +1415,17 @@ service AuthService {
...
+    // GetMFAAuthenticateChallengeWithToken retrieves challenges for all mfa devices for the user
     // defined in the token.
     rpc GetMFAAuthenticateChallengeWithToken(GetMFAAuthenticateChallengeWithTokenRequest)
         returns (MFAAuthenticateChallenge);
     // GetMFADevicesWithToken returns all mfa devices for the user defined in the token.
     rpc GetMFADevicesWithToken(GetMFADevicesWithTokenRequest) returns (GetMFADevicesResponse);
     // DeleteMFADeviceWithToken deletes a mfa device for the user defined in the token.
     rpc DeleteMFADeviceWithToken(DeleteMFADeviceWithTokenRequest) returns (google.protobuf.Empty);
+    // AddMFADeviceWithToken adds a mfa device for the user defined in the token.
+    rpc AddMFADeviceWithToken(AddMFADeviceWithTokenRequest) returns (google.protobuf.Empty);
```
Comment (codingllama) [inline api/client/proto/authservice.proto:1430]: There is a lot of API surface for MFA already, but it seems we're effectively doubling it by adding new `WithToken` methods for the lot of it.

Have we considered adding this functionality on top of the existing RPCs? We could support privilege tokens as a way of authenticating in the streaming methods, cutting down the necessary interactions to the bare minimum. Let's talk about this before you go much further in the implementation, as what we are doing is going to impact each other's work.

File: api/client/proto/authservice.proto [hunk 11]
```diff
@@ -1053,6 +1073,19 @@ message CreateRecoveryCodesWithTokenResponse {
     repeated string RecoveryCodes = 1;
 }
 
+// CreatePrivilegeTokenRequest defines a request to create and get a user privilege token.
```
Comment (codingllama) [inline api/client/proto/authservice.proto:1076]: I assume the privilege token exists so the user doesn't have to keep tapping the existing token in a small time window, correct?

Please explain why someone would create a token, what it does and how long it's expected to last.

File: api/client/client.go (hunk 2 above)
Comment (codingllama) [reply api/client/client.go:1150]: +1

File: lib/auth/auth.go [hunk 12]
```diff
@@ -1214,6 +1225,92 @@ func (a *Server) DeleteMFADeviceWithToken(ctx context.Context, req *proto.Delete
...
+			return trace.BadParameter("cluster only allows OTP for second factor")
+		}
+
+		cfg, err := authPref.GetU2F()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		device, err := a.createNewU2FDevice(ctx, newU2FDeviceRequest{
```
Comment (codingllama) [inline lib/auth/auth.go:1263]: Device creation only happened at addMFADeviceRegisterChallenge - I don't think we should be handling this in 2 places. This is only getting more complex with the addition of Webauthn.

File: lib/auth/auth.go (hunk 6 above)
Comment (codingllama) [reply lib/auth/auth.go:1254]: It's strange for an user to reply to an U2F challenge the cluster didn't supply - they shouldn't have the challenge and it'll never pass anyway.

I guess this could happen for OTP, so maybe let's handle only that. As stated in other comments, though, I think a single method should handle this (it used be addMFADeviceRegisterChallenge).

File: api/client/client.go [hunk 13]
```diff
@@ -1137,6 +1147,16 @@ func (c *Client) GetMFADevices(ctx context.Context, in *proto.GetMFADevicesReque
 	return resp, nil
 }
 
+// GetMFAAuthenticateChallengeWithAuth gets mfa challenges for the currently logged in user.
+func (c *Client) GetMFAAuthenticateChallengeWithAuth(ctx context.Context, in *proto.GetMFAAuthenticateChallengeWithAuthRequest) (*proto.MFAAuthenticateChallenge, error) {
+	resp, err := c.grpc.GetMFAAuthenticateChallengeWithAuth(ctx, in, c.callOpts...)
+	if err != nil {
```
Comment (codingllama) [inline api/client/client.go:1153]: Ditto, drop the if and do a single return. Same throughout, where applicable.

File: lib/auth/usertoken.go [hunk 14]
```diff
@@ -52,6 +55,10 @@ const (
 	// This token is also used to allow users to delete exisiting second factor devices
 	// and retrieve their new set of recovery codes as part of the recovery flow.
 	UserTokenTypeRecoveryApproved = "recovery_approved"
+	// UserTokenTypePrivilege describes a token type that grants access to a privileged action
+	// that requires users to re-authenticate with their second factor while looged in. This
```
Comment (codingllama) [inline lib/auth/usertoken.go:59]:
```suggestion
	// that requires users to re-authenticate with their second factor while logged in. This
```

File: lib/auth/usertoken.go [hunk 15]
```diff
@@ -510,3 +520,130 @@ func (s *Server) getRecoveryStartToken(ctx context.Context, tokenID string) (typ
...
+	}
+
+	secondFactor := authPref.GetSecondFactor()
+	if secondFactor == constants.SecondFactorOff {
+		return nil, trace.AccessDenied("second factor must be enabled")
+	}
+
+	// Begin authenticating user password or second factor.
+	switch req.GetSecondFactorCred().(type) {
+	case *proto.CreatePrivilegeTokenRequest_SecondFactorToken:
```
Comment (codingllama) [inline lib/auth/usertoken.go:542]: Let's not duplicate MFA handling code; let's find a way to integrate this into the existing logic.

**Review by codingllama — CHANGES_REQUESTED** [review]
I'm glad you copied me here because it looks like we need to sync.

I'm currently working on adding WebAuthn as an additional MFA protocol, eventually replacing U2F. Next on the pipeline is moving login to gRPC, which touches this even further. The bulk of my suggestions, then, are around keeping API surface small and make code duplication related to MFA close to inexistent.

Let me know if you wan to chat.

**Comment by codingllama** [comment]
Just had a chat with @kimlisa, so I'll try to sum up here what we discussed.

1. Ideally, IMO, we'd move the Web UI to gRPC-web and consume the existing streaming RPCs for MFA. (Likely keeping the new user token methods and adding support for it in existing challenges.)

2. The work necessary for gRPC-web support might impede (1), in which case we'd have non-streaming versions of MFA RPCs, but with the full range of features and code flowing to a single point to avoid duplication. If the expectation for gRPC-web is very low, we might go as far as removing the streaming versions instead (which I would prefer instead of having duplicated API surface).

Either way, ideally, we'd have a small RFD delineating the changes.

This goes way beyond the scope of this PR. There is also a tight deadline for MFA UI support, which I don't want to jeopardize, so we decided to ask Sasha for input - we may go ahead and get the feature out, with the understanding that a refactor would come into play later.

FYI @klizhentas @alex-kovoy @russjones

File: api/client/proto/authservice.proto (hunk 10 above)
Comment (codingllama) [reply api/client/proto/authservice.proto:1430]: Resolving based on various conversations we had.

**Comment by kimlisa** [comment]
closing with new PR: https://github.com/gravitational/teleport/pull/8067
